### ISSUE TITLE
Improve performance (also by using hardware acceleration)

### ISSUE THREAD

**Issue by hawkerboy7** [bevacqua/dragula#303]
I am creating an application with a lot of content shown/rendered on the front-end and I am noticing performance issues in Dragula.

Dragula appears to keeps setting a `class` and `styles` on `onmousemove`.
This causes re-paints in the browser which is bad for performance.
I might be analyzing the problem incorrectly so please correct me if I am wrong but I see the following things slowing down performance.

`onmousemove` (so on every mouse move (which is quite a lot!) the following changes/is set):
- `class gu-mirror` is being set (even though it's only required once)
- `width` and `height` are set (which is only required at the start of dragging and when the window resizes)
- position `top` and `left` are updated (which should be turned into `transform: translateZ(0px) translate( X px, Y px)`)

The `class` being set should, I think, only be done when the drag starts. After that you don't have to set it again (especially not if it's the exact same class).

The `width` and `height` should also only be set when the drag starts. They however may be updated when the `window` resizes (this isn't occurring now but could be a nice addition/new feature).

Position `top` and `left` on every mousemove causes a re-paint for each mousemove.
Changing this into `transform: translateZ(0px) translate( X px, Y px)` will trigger hardware acceleration. This provides a better performance because re-paint don't occur.

I'd appreciate your view on these improvements and if I can help writing some code I'd like to hear it and I'll try to create some free time

**Comment by michaelmalonenz** [comment]
I'm with you on everything except for the specific shape of the transform css:  

Internet Explorer doesn't support 3D transforms (and we don't need them) so let's just do a `transform: translate( X px, Y px);`  (so if it's the z-index that triggers the hardware accel, then we might have to make it an opt-in thing)

**Comment by luckylooke** [comment]
How do you want to pass mouse coords to translate( X px, Y px)?

**Comment by hawkerboy7** [comment]
@luckylooke I don' t think I understand your question. Now the mouse coords go to `top` and `left` so you just change it into `translate`.

```
el.style.top = Y + ' px' 
el.style.left = X + ' px'
```
becomes
```
el.style.transform = 'translateZ(0)translate('+X+'px,'+Y+'px)'
```


@michaelmalonenz You could use something else to trigger the hardware acceleration, what do you have in mind? As far as I know `z-index` doesn't trigger it.

**Comment by luckylooke** [comment]
@hawkerboy7 you got it right.. thanks.. do you have some reference that it is faster solution?

**Comment by hawkerboy7** [comment]
There are [loads](http://lmgtfy.com/?q=hardware+acceleration+css3+transform+performance) of references on google...

These are two for example:
http://www.html5rocks.com/en/tutorials/speed/layers/
http://www.paulirish.com/2012/why-moving-elements-with-translate-is-better-than-posabs-topleft/

**Comment by luckylooke** [comment]
@hawkerboy7 all use cases I have found, including those you provide are transforms hardcoded in css not transforms updated every mousemove, thats huge difference in my opinion. I understand performance gain using transform for animations but not for following mouse cursor. Do you have reference that someone somewhere noticed performance gain using transform in dragging?

**Comment by bevacqua** [comment]
There was a PR for this at some point (#210), but it relied on an option rather than feature detection so I never merged it. I'd be okay merging something like this if `(a)` we figure out if it's noticeably better as @luckylooke says, and `(b)` it relies on feature-detection as opposed to an option flag.

**Comment by hawkerboy7** [comment]
@luckylooke I think either I am missing your point or you are missing mine :) I don't fully understand your sentences.

On the event `mousemove` which is triggered every time your mouse is moved `dragula` currently changes the `el.style.top/left` you could say that's 'hardcoded in css' if you like, since css properties change ...
But now you change `el.style.transform` instead which then also is 'hardcoded in css' ...

How is an animation different from a position being set on an element?
If you use `jQuery.animate()` your `top` and `left` properties are changed rapidly to create the effect of an animation. Changing the `top` and `left` property by making those values the exactly position of your mouse every time the mouse moves isn't any different.

The performance gain you get as I said before is because any animation or mouse 'follow' action now triggers a repaint in the browser. The bigger and the more css properties the element your dragging has the more you'll notice performance getting worse (initially on smaller elements it doesn't have a bad performance, it works fine then)

To check:
- just open http://bevacqua.github.io/dragula/ in chrome.
- open the inspector
- open the console on the bottom of the inspector (using escape, click it a few times to see it appear and disappear)
- in the options you'll find 'Rendering'
- press the checkbox of 'Enable paint flash rendering'
- Now drag any element and you'll see a green box around the element, which indicates a repaint.

**Comment by bevacqua** [comment]
@hawkerboy7 are you interested at all in implementing a PR for this?

**Comment by hawkerboy7** [comment]
@bevacqua (a) It's noticeable if the elements you're dragging contain a lot of css and or html elements.
(b) the cool thing about css3 is that if hardware acceleration works it does it and if it doesn't it just works as normal so you won't have to do feature detection, unless I am missing something?
Which browsers do you want to support? All but < IE 11

**Comment by hawkerboy7** [comment]
@bevacqua Yes I could make one, however I don't want to spend time in code which is not going to be implemented as you can understand :) I'd like to know your objections against it, so I can determine whether or not to start working on it.

**Comment by romainbessugesmeusy** [comment]
Hey, sorry to interrupt, but the #210 you just closed has been updated for a while and relies on feature detection and not options anymore:
https://github.com/bevacqua/dragula/commit/2c24df3d9fa712f55b7328de6a3669bfe62a3534

**Comment by luckylooke** [comment]
@hawkerboy7 I apologize for my English, I am not native speaker and I am doing my best to be clear as much as possible.

By 'hardcoded' I was pointing to references you sent. All of them uses defined transforms in css styles and then just starting them by adding/removing classes on element. So there is missing test if dynamicaly setting transform every mousemove have expected performance gain. I dont claim it doesn't have one, I was asking for reference. That was all.

If you are sure, go ahead and implement it, I am looking forward to your PR. ;)

**Comment by hawkerboy7** [comment]
@luckylooke You don't add/remove classes on an element.  You just adjust the style, just like what's happening right now. I have not made a test indeed I also don't have a reference in which you can see dragula performing better because the solution has not been made yet. If you are refering to an actual test in which you can see that hardwareacceleration actually improves performance, you can easily find those on google.

Ok good :)

**Comment by romainbessugesmeusy** [comment]
Hey Nicolás, would you please take another look to my previous PR ? It does not rely on options anymore...

**Comment by xiwc** [comment]
I also met the same problem: drag and drop is not smooth, when the data is more. 顶起!

**Comment by swederik** [comment]
For what it's worth, @romainbessugesmeusy's PR #210 to use transform instead of top / left works very well. I only had a one tiny issue: the feature detection uses [remove()](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/remove), which is not supported below IE11. 

Romain's useTransform branch needed a rebase as well. An updated version is [here](https://github.com/swederik/dragula/tree/useTransform) if anyone needs it. I could open a PR but I don't want to take the credit for @romainbessugesmeusy's nice work!

Thanks for a very nice library!

**Comment by limrs** [comment]
Hi @hawkerboy7, is it possible to remove paint flash rendering completely by using css only/ making the list of item become css 3d without modifying the dragula javascript

**Comment by romainbessugesmeusy** [comment]
From my experience, it does not suffice. The real pain for the browser is the absolute positioning using top and left using floats. Adding transform3d is just a trick to use in combination with translate to tell the browser to do a hardware rendering instead of software.

**Comment by hawkerboy7** [comment]
Hi @limrs. If I understand you correctly you want to use css3 hardware acceleration to create a new composite layer but still use `.top/left` to adjust the position? Hereby preventing a re-paint? You can try this on the example page of dragula by adding the following property to the `gu-miror` : `will-change: top,left;`. You'll see this doesn't work.

This creates the new layer (so it's hardware accelerated) but if you still use existing dragula code you'd still be changing the top and left. Eventhough it's in a new layer this will still cause a repaint because chaning top/left always triggers a paint :(

**Comment by limrs** [comment]
Yes, @hawkerboy7 , that is what i try to achieve. My Colleagues add the following css to those elmement that will be drag 

` -webkit-transform: translateZ(0);
    -moz-transform: translateZ(0);
    -ms-transform: translateZ(0);
    -o-transform: translateZ(0);
    transform: translateZ(0);
    -webkit-backface-visibility: hidden;
    -webkit-perspective: 1000;` 

it seem like it temporarily solve my issue. and thanks for hepling me :)

**Comment by hawkerboy7** [comment]
@limrs Good! Glad you figured out how to improve the performance, maybe using that hardware acceleration doesn't stop the re-paint of the parent div (the one that's dragged) but it may stop re-paints of the content of that div which might explain your performance gain. Good to know thx :)

**Comment by hawkerboy7** [comment]
@luckylooke @bevacqua 

Hi guys,

It's been a while. 

I noticed in my application as feedback that some users wanted to click an item but instead started to drag it. Also there where some performance issues (on IE) when they started dragging a lot of content around that I mentioned before in this thread. 

As a result I tried to create my own version of a drag and drop library that would support these problems. It is not as rich in features and support as Dragula, but I've added one feature which you may be very interested in for implementing in Dragula as wel!
It's called [Threshold](https://github.com/hawkerboy7/dragify#optionsthreshold).
As the description states this will help users that can't keep their mouse steady during the click and accidentally start dragging instead of clicking an element.
Also the hardware acceleration using css3 transform and will-change has been applied and appears to work well.

If you've got time maybe take a look. It would be great if Dragula could work in a similar way :)
[Dragify](https://github.com/hawkerboy7/dragify)

**Comment by ibrychgo** [comment]
limrs, thank you. Immediate performance gain. The jitteryness went away.

**Comment by lloyd-2014** [comment]
@limrs I LOVE YOU
